feat: Add ViewModel in SignUpActivity

dr0pdb opened the pull request:
Fixes #2123 

Changes: Added ViewModel class for SignUpActivity

Screenshots for the change: NA

@iamareebjamal Please review.

iamareebjamal reviewed (changes requested):
Move the actual API call in ViewModel

dr0pdb commented:
Changes: Moved the actual API call to the ViewModel
Since the ```loginUser``` method inside AuthUtils is also used by the LoginActivity, hence i couldn't edit it. So i created another method ```loginUserWithoutProgressBar```. It will be renamed to ```loginUser``` and the actual ```loginUser``` method will be deleted.
Since the ```signUpUser``` method was only used by the SignUpActivity hence i modified it.

GIF:
![videotogif_2018 01 22_15 09 42](https://user-images.githubusercontent.com/21277837/35214827-7694af1a-ff88-11e7-83de-30ea64e87426.gif)


@iamareebjamal Please review.

simarsingh24 commented on android/app/src/main/java/org/fossasia/openevent/activities/auth/SignUpActivity.java:
```diff
@@ -61,6 +64,8 @@ protected void onCreate(Bundle savedInstanceState) {
         confirmPasswordInput = (TextInputEditText) confirmPasswordWrapper.getEditText();
 
+        signUpActivityViewModel = ViewModelProviders.of(this).get(SignUpActivityViewModel.class);
+
```
Please remove this space

dr0pdb replied:
@simarsingh24 The aim is to make things clearer to the reader. If you observe the whole method closely(the parts that are left untouched in this PR), you will realise that spaces have been provided between unrelated things. For eg:

```Java

setSupportActionBar(toolbar);
getSupportActionBar().setDisplayHomeAsUpEnabled(true);

switchToLogin.setOnClickListener(v -> startActivity(new Intent(SignUpActivity.this, LoginActivity.class)));
```

So to remain consistent with what has been done, i think it's better to add a space as that line is unrelated to anything above/below it.

simarsingh24 replied:
if that's the idea then its fine :)

simarsingh24 commented:
Why don't you use a boolean for progress bar inside` loginUser` method of AuthUtils and if it is being used by LoginActivity why not considering making some changes there also otherwise it would cause duplication of code!? @srv-twry what do you think?

simarsingh24 commented on android/app/src/main/java/org/fossasia/openevent/viewmodels/auth/SignUpActivityViewModel.java:
```diff
@@ -0,0 +1,60 @@
+import android.arch.lifecycle.LiveData;
+import android.arch.lifecycle.MutableLiveData;
+import android.arch.lifecycle.ViewModel;
+
```
remove this space please

dr0pdb replied:
This space is automatically provided by Android studio. Android studio automatically groups the similar imports as one and adds spaces between unrelated imports .It's suitable not to change it.

dr0pdb commented:
@simarsingh24 

> Why don't you use a boolean for progress bar insideloginUser method of AuthUtils 

Pardon, didn't get you. Can you please elaborate.

> if it is being used by LoginActivity why not considering making some changes there also otherwise it would cause duplication of code!?

Because that is not the aim of this PR. Editing LoginActivity interaction should be done in a separate PR. Infact since we will adding a ViewModel for LoginActivity, it's better to refactor it then.

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/utils/AuthUtil.java:
```diff
@@ -34,23 +36,45 @@
                             showMessage(R.string.signed_up_successfully);
                             Timber.d("Signed up successfully");
+                            signUpResponse.setValue(SignUpActivityViewModel.VALID);
+                        });
```
Since this observable emits a single, completed handler is same as onNext handler, move this up to `user -> {}`

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/utils/AuthUtil.java:
```diff
@@ -34,23 +36,45 @@
+    }
+
+    public static void loginUserWithoutProgressBar(String email, String password, MutableLiveData<Integer> loginResponse) {
+        APIClient.getOpenEventAPI().login(new Login(email, password))
```
Add to composite disposable

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/utils/AuthUtil.java:
```diff
@@ -34,23 +36,45 @@
+                        response -> {
+                            //Save token & email in shared preferences
+                            SharedPreferencesUtil.putString(ConstantStrings.TOKEN, response.getAccessToken());
+                            SharedPreferencesUtil.putString(ConstantStrings.USER_EMAIL, email);
```
Don't use Android-specific classes in ViewModels

dr0pdb replied:
It is using the Application context which i think shouldn't be a problem. Any other reason why we should avoid it in ViewModels ?
I'll move it to the view though.

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/utils/AuthUtil.java:
```diff
@@ -34,23 +36,45 @@
     private static Authenticator authenticator;
 
-    public static void signUpUser(Context context, String email, String password, ProgressBar progressBar) {
+    public static void signUpUser(String email, String password, MutableLiveData<Integer> signUpResponse) {
```
The idea is to deconstruct AuthUtil class, move this whole method inside the signup viewmodel. I mean there is no point of hoisting a single use method in a different class

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/viewmodels/auth/SignUpActivityViewModel.java:
```diff
@@ -0,0 +1,60 @@
+        if (loginResponse == null) {
+            loginResponse = new MutableLiveData<>();
+        }
+        AuthUtil.loginUserWithoutProgressBar(email, password, loginResponse);
```
If this is not used anywhere else. Move this in here please

dr0pdb replied:
It will be used in the future by LoginActivity. 
I had to refactor both the signUp and loginUser methods. signUp user method was only used by SignUpActivity so i did it. The loginUser method was also used by the LoginActivity, hence i couldn't afford editing that. Hence i created a temporary method loginUserWithoutProgressBar to serve the purpose.
The idea is to remove the loginUser method and rename this loginUserWithoutProgressBar method to loginUser and make the LoginActivity use this method using it's ViewModel once it is created.

dr0pdb commented:
Changes: Moved signUpUser code to ViewModel, Using composite disposable, Using Android-specific classes inside the View, Moved onComplete handler code to onNext.

@iamareebjamal Please review.

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/utils/AuthUtil.java:
```diff
@@ -34,23 +37,22 @@
-                        disposable -> showProgressBar(progressBar, true));
+                            loginResponse.setValue(new Pair<>(SignUpActivityViewModel.INVALID, ""));
+                        }));
+        Timber.d(email);
```
Please remove

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/utils/AuthUtil.java:
```diff
@@ -34,23 +37,22 @@
                 .observeOn(AndroidSchedulers.mainThread())
                 .subscribe(
-                        user -> loginUser(context, email, password, progressBar),
+                        response -> {
```
Move to previous line

dr0pdb replied:
Sorry, didn't get you. Please explain a bit more.

iamareebjamal replied:
response -> can be moved to previous line, i.e., subscribe(response ->

iamareebjamal commented on android/app/src/main/java/org/fossasia/openevent/utils/AuthUtil.java:
```diff
@@ -34,23 +37,22 @@
+    public static void loginUserWithoutProgressBar(String email, String password,
+                                                   MutableLiveData<Pair<Integer,String>> loginResponse,
+                                                   CompositeDisposable compositeDisposable) {
+        compositeDisposable.add(APIClient.getOpenEventAPI().login(new Login(email, password))
```
Please move this entire code to ViewModel itself. Is this code used anywhere else as well?

dr0pdb replied:
No, but it will be used by the LoginActivity when it's ViewModel is implemented.

iamareebjamal replied:
For now, move it into ViewModel because it is simply taking the environment of ViewModel and passing into the arguments. We'll think of something better when there is time for sharing code

dr0pdb committed:
feat: Add ViewModel in SignUpActivity

dr0pdb commented:
Changes: Moved loginUser code to ViewModel and minor formatting changes. Removed logging of email
@iamareebjamal Please review.
